source/common/network/connection_impl.cc

```diff
@@ -645,15 +645,18 @@ void ConnectionImpl::onWriteReady() {
   } else if ((inDelayedClose() && new_buffer_size == 0) || bothSidesHalfClosed()) {
     ENVOY_CONN_LOG(debug, "write flush complete", *this);
     if (delayed_close_state_ == DelayedCloseState::CloseAfterFlushAndWait) {
-      ASSERT(delayed_close_timer_ != nullptr);
-      delayed_close_timer_->enableTimer(delayed_close_timeout_);
+      ASSERT(delayed_close_timer_ != nullptr && delayed_close_timer_->enabled());
+      if (result.bytes_processed_ > 0) {
+        delayed_close_timer_->enableTimer(delayed_close_timeout_);
```

Comment:
I don't think there's a way to distinguish between these two enableTimer calls in tests. Ideally we would move delayed_close_state_ to a WaitingForClose state or change the fd mask to:
  file_event_->setEnabled(enable_half_close_ ? 0 : Event::FileReadyType::Closed);
to avoid getting any Write events past this point.

Reply:
nah, but I'll trust you made sure your test would fail without the code change :-)